Utilities/XrdAdaptor/src/XrdFile.cc

```diff
@@ -248,9 +250,54 @@ XrdFile::read (void *into, IOSize n, IOOffset pos)
     addConnection(ex);
     throw ex;
   }
+  if (n == 0) {
+    return 0;
+  }
+
+  // In some cases, the IO layers above us (particularly, if lazy-download is
+  // enabled) will emit very large reads.  We break this up into multiple
+  // reads in order to avoid hitting timeouts.
+  std::vector<IOPosBuffer> requests;
+  while (n) {
+    IOSize chunk = std::min(n, static_cast<IOSize>(XRD_CL_MAX_READ_SIZE));
+    requests.emplace_back(pos, into, chunk);
+    into = static_cast<char*>(into) + chunk;
+    n -= chunk;
+    pos += chunk;
+  }
 
-  uint32_t bytesRead = m_requestmanager->handle(into, n, pos).get();
+  std::vector<std::pair<std::future<IOSize>, IOSize>> futures; futures.reserve(requests.size());
+  futures.emplace_back(m_requestmanager->handle(requests[0].data(), requests[0].size(), requests[0].offset()),
+                       requests[0].size());
+  for (unsigned idx = 0; idx < requests.size(); idx++) {
+    if (idx + 1 < requests.size()) {
+      futures.emplace_back(m_requestmanager->handle(requests[idx+1].data(),
+                                                    requests[idx+1].size(),
+                                                    requests[idx+1].offset()),
+                           requests[idx+1].size());
+    }
+    if (futures[idx].first.valid()) {
+      futures[idx].first.wait();
+    }
+  }
 
+  IOSize bytesRead = 0;
+  bool readReturnedShort = false;
+  for (auto &future : futures) {
+    // Future throws an exception on failure.
+    IOSize result = future.first.get();
```

Comment:
Ok, I think I came up with a clean / readable way to do this without maintaining a list.  Will update in a moment.